tortoise/contrib/pydantic/dataclasses.py

```diff
@@ -0,0 +1,264 @@
+    model_name: str = ""
+    related_name: str = ""
+    forward_key: str = ""
+    backward_key: str = ""
+    through: str = ""
+    on_delete: str = ""
+    _generated: bool = False
+
+
+@dataclasses.dataclass
+class ModelDescription:
+    name: str
+    table: str
+    abstract: bool
+    description: Optional[str]
+    pk_field: FieldDescriptionBase
+    app: Optional[str] = None
+    docstring: Optional[str] = None
+    unique_together: Tuple[Tuple[str, ...], ...] = dataclasses.field(default_factory=tuple)
+    indexes: Tuple[Tuple[str, ...], ...] = dataclasses.field(default_factory=tuple)
+    data_fields: List[FieldDescriptionBase] = dataclasses.field(default_factory=list)
+    fk_fields: List[FieldDescriptionBase] = dataclasses.field(default_factory=list)
+    backward_fk_fields: List[FieldDescriptionBase] = dataclasses.field(default_factory=list)
+    o2o_fields: List[FieldDescriptionBase] = dataclasses.field(default_factory=list)
+    backward_o2o_fields: List[FieldDescriptionBase] = dataclasses.field(default_factory=list)
+    m2m_fields: List[FieldDescriptionBase] = dataclasses.field(default_factory=list)
+
+
+def describe_model_by_dataclass(cls: Type[MODEL]) -> ModelDescription:
+    return ModelDescription(
+        name=cls._meta.full_name,
+        app=cls._meta.app,
+        table=cls._meta.db_table,
+        abstract=cls._meta.abstract,
+        description=cls._meta.table_description or None,
+        docstring=inspect.cleandoc(cls.__doc__ or "") or None,
+        unique_together=cls._meta.unique_together or (),
+        indexes=cls._meta.indexes or (),
+        pk_field=describe_field_by_dataclass(cls._meta.fields_map[cls._meta.pk_attr]),
+        data_fields=[
+            describe_field_by_dataclass(field)
+            for name, field in cls._meta.fields_map.items()
+            if name != cls._meta.pk_attr and name in (cls._meta.fields - cls._meta.fetch_fields)
+        ],
+        fk_fields=[
+            describe_field_by_dataclass(field)
+            for name, field in cls._meta.fields_map.items()
+            if name in cls._meta.fk_fields
+        ],
+        backward_fk_fields=[
+            describe_field_by_dataclass(field)
+            for name, field in cls._meta.fields_map.items()
+            if name in cls._meta.backward_fk_fields
+        ],
+        o2o_fields=[
+            describe_field_by_dataclass(field)
+            for name, field in cls._meta.fields_map.items()
+            if name in cls._meta.o2o_fields
+        ],
+        backward_o2o_fields=[
+            describe_field_by_dataclass(field)
+            for name, field in cls._meta.fields_map.items()
+            if name in cls._meta.backward_o2o_fields
+        ],
+        m2m_fields=[
+            describe_field_by_dataclass(field)
+            for name, field in cls._meta.fields_map.items()
+            if name in cls._meta.m2m_fields
+        ],
+    )
+
+
+def describe_field_by_dataclass(field: Field) -> FieldDescriptionBase:
```

Comment:
Could you please help me to understand what is the benefit of converting the model and field metadata to a dataclass, so then we can convert it to a pydantic model? Not a criticism! Just trying to understand the logic.

Reply:
before using dataclasses, everything was inside a simple dict. A dict does not provide any typing information, and you also do not know what is inside a dict, with other words: you have no control about what keys are in a dict and also have no clue about of what type the values are. Since in this case it is very predictable what is needed for pydantic_model_creator, I found it much more consistent to rely on dataclasses, that do exactly what dicts don't. In my opinion, it is a good compromise between light weight and structural integrity.

Reply:
another approach would be to not use dataclasses or a dict and directly read the needed values from the Field Instances

Reply:
> another approach would be to not use dataclasses or a dict and directly read the needed values from the Field Instances

that's not possible because Field.constraints is modified during the process.